### PR TITLE
feat(coachmark): coachmark component

### PR DESCRIPTION
**Pull request by helen-v**
## Screenshots:
**Before** (If applicable):

**After**:

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: web-components/src/components/coachmark/Coachmark.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      super.handleFocusIn(event);
+    }
+    this.notifyCoachCreate();
+  }
+
+  protected handleFocusOut(event: Event) {
+    if (super.handleFocusOut) {
+      super.handleFocusOut(event);
+    }
+  }
```
Comment (boiarchukk) [inline web-components/src/components/coachmark/Coachmark.ts:41]: @helen-v u can easily delete this code

File: web-components/src/components/coachmark/Coachmark.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      if (slotContent.length) {
+        this.slotContent = slotContent;
+      }
+    }
+  }
+
+  get coachClassMap() {
+    return {
+      [`md-coachmark__popper--${this.placement}`]: this.placement,
+      [`md-coachmark__popper--${this.color}`]: this.color
```
Comment (boiarchukk) [inline web-components/src/components/coachmark/Coachmark.ts:70]: Use need convert to boolean value:
` [`md-coachmark__popper--${this.placement}`]: !!this.placement,
  [`md-coachmark__popper--${this.color}`]: !!this.color`

File: web-components/src/components/coachmark/Coachmark.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    };
+  }
+
+  render() {
+    return html`
+      <div class="md-coachmark ${this.show ? 'md-coachmark--active' : ''}">
```
Comment (boiarchukk) [inline web-components/src/components/coachmark/Coachmark.ts:76]: Please use [classMap](https://lit-html.polymer-project.org/guide/template-reference#classmap) directive

File: web-components/src/components/coachmark/Coachmark.ts (hunk 1 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.ts:41]: Removed

File: web-components/src/components/coachmark/Coachmark.ts (hunk 2 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.ts:70]: Done

File: web-components/src/components/coachmark/Coachmark.ts (hunk 3 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.ts:76]: Done

File: web-components/src/components/coachmark/Coachmark.test.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    const wrap = component.shadowRoot?.querySelector(".md-coachmark");
+    expect(wrap?.getAttribute("class")).toEqual("md-coachmark md-coachmark--active");
+  });
+
+  test("should dispatch Action", async () => {
+    const component = await fixtureFactory();
+
+    const mockClick = jest.spyOn(component, "coachAction");
+    component.coachAction(new CustomEvent("click"));
```
Comment (futurethang) [inline web-components/src/components/coachmark/Coachmark.test.ts:40]: I'm getting an error that prevents compile on this line of the test

File: web-components/src/components/coachmark/Coachmark.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    expect(mockClick).toHaveBeenCalled();
+
+    mockClick.mockRestore();
+  });
+
+  test("should notify Coachmarkv create", async () => {
+    const component = await fixtureFactory();
+
+    const mockCreate = jest.spyOn(component, "notifyCoachCreate");
+    component.notifyCoachCreate(new CustomEvent("click"));
```
Comment (futurethang) [inline web-components/src/components/coachmark/Coachmark.test.ts:52]: I'm getting an error that prevents compile on this line of the test

File: web-components/src/components/coachmark/Coachmark.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+          <slot></slot>
+        </div>
+      </div>
+    `;
+  }
+
+  static get styles() {
+    return [reset, styles];
+  }
+}
```
Comment (futurethang) [inline web-components/src/components/coachmark/Coachmark.ts:129]: Nit: I think it works either way, but this style loader is usually placed just before the render method in most other components, and that seems like a nice location for it to consistently be.

**Review by futurethang — CHANGES_REQUESTED** [review]
When a new coachmark opens, the focus falls on the popup box, requiring a tab-in to the 'next' button. Is that the correct pattern? I feel like focusing on the 'next' button is more natural so a user can keep hitting enter or space to get to the next tip.

Otherwise looking great!

File: web-components/src/components/coachmark/Coachmark.ts (hunk 6 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.ts:129]: Agree. Fixed.

File: web-components/src/components/coachmark/Coachmark.test.ts (hunk 5 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.test.ts:52]: Fixed

File: web-components/src/components/coachmark/Coachmark.test.ts (hunk 4 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.test.ts:40]: fixed

**Comment by helen-v** [comment]
In this case, this was done to determine the focus on the internal elements of the component, since inside there can be a slot with other elements. In my opinion, this is the most optimal option with focus.

File: web-components/src/components/tooltip/scss/tooltip.scss [hunk 7]
```diff
@@ -32,7 +32,7 @@
       text-align: center;
       width: max-content;
       word-break: break-word;
-      z-index: -1;
+      z-index: 9999;
```
Comment (yananym) [inline web-components/src/components/tooltip/scss/tooltip.scss:35]: Lena you can't simply put 9999. Please evaluate other z-indexs. there is a hierarchy here: https://sqbu-github.cisco.com/CBABU/agentx-services/blob/master/src/styles/z-index.scss

File: web-components/src/components/coachmark/Coachmark.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+  static get styles() {
+    return [reset, styles];
+  }
+
+  render() {
+    return html`
+      <div class="md-coachmark ${classMap(this.coachWrapClassMap)}">
+        <div class="md-coachmark__popper ${classMap(this.coachClassMap)}" tabindex="0">
+          <div class="md-coachmark__content" >
```
Comment (yananym) [inline web-components/src/components/coachmark/Coachmark.ts:105]: Doesn't seem like you are actually using popper.js here?

File: web-components/src/components/tooltip/scss/tooltip.scss (hunk 7 above)
Comment (helen-v) [reply web-components/src/components/tooltip/scss/tooltip.scss]: Fixed

File: web-components/src/components/coachmark/Coachmark.ts (hunk 8 above)
Comment (helen-v) [reply web-components/src/components/coachmark/Coachmark.ts:105]: Not sure if we need here popper? If yes, I fix it.

**Review by futurethang — APPROVED** [review]

**Pull request merged**